### DISCUSSION TITLE
Examples of functions that stay at a single level of abstraction

### DISCUSSION THREAD

**Discussion by maurolepore** [2DegreesInvesting/practices#10] [Show and tell]
> It turns out to be very difficult ... to write functions that stay at a single level of abstraction. [But] it is the key to keeping functions short and making sure they do "one thing". -- Bob Martin (Clean Code)

Because this is so important but hard to do, I would like to collect here some concrete examples in R of functions that do this well. I'll add examples as I come across them, and I encourage others to do the same.

Thanks @cjyetman.

### Reference 

![IMG-0081](https://user-images.githubusercontent.com/5856545/105427031-e1f41000-5c11-11eb-8c05-4678c34e34a1.jpg)

![IMG-0082](https://user-images.githubusercontent.com/5856545/105427024-de608900-5c11-11eb-8621-59ae1d17d857.jpg)

**Comment by jdhoffa** [comment]
I would welcome a `ds-incubator` or `dev-hangout` explaining the "Stepdown Rule" with some R code examples! I will try to think of some as I work today.

**Comment by maurolepore** [comment]
Here is one real-life example where I made a concious effort to apply the stepdown rule. My goal here is not to argue this is good code but to clarify with a concrete example how I interpret the stepdown rule:

```r
use_bridge <- function(dataset,
                       data,
                       contributor = NULL,
                       issue = NULL,
                       overwrite = FALSE) {
  dataset <- bridge_name(dataset)
  bridge_write_raw_data(dataset, data, overwrite = overwrite)
  bridge_add_dictionary(dataset, overwrite = overwrite)
  bridge_update_news(dataset, contributor = contributor, issue = issue)
  bridge_update_r(dataset)
  bridge_update_data_raw(dataset)

  refresh_data()
  message(bridge_todo())

  invisible(dataset)
}

bridge_name <- function(prefix) {
  prefix <- tolower(prefix)
  has_underscore <- grepl("_", prefix)

  if (has_underscore) {
    stop(
      "`prefix` '", prefix, "' must not contain unterscore '_'.",
      call. = FALSE
    )
  }

  glue("{prefix}_classification")
}

# ... many more low level functions
```

Notice that changes in low level details (like the case of the dataset name or the error message) would not affect the high level function at all. (The high level function it would only change due to significant changes in the inputs of the lower level functions).
